File: src/main/java/hobby/wei/c/reflow/In.java
/*
 * Copyright (C) 2016-present, Wei.Chou(weichou2010@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package hobby.wei.c.reflow;

import java.lang.ref.WeakReference;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import hobby.wei.c.tools.Locker;

import static hobby.wei.c.reflow.Assist.*;
import static hobby.wei.c.reflow.Assist.getRef;
import static hobby.wei.c.reflow.Assist.requireKey$kDiff;
import static hobby.wei.c.reflow.Assist.requireNonNullElem;
import static hobby.wei.c.reflow.Assist.requireTransInTypeSame;

/**
 * @author Wei.Chou (weichou2010@gmail.com)
 * @version 1.0, 14/08/2016
 */
public abstract class In {
    private static WeakReference<In> sEmptyRef;

    final Set<Key$> keys;
    Set<Transformer> trans;

    protected In(Set<Key$> keys) {
        this.keys = requireKey$kDiff(requireNonNullElem(keys));
    }

    public In transition(Transformer trans) {
        return transition(Collections.singleton(trans));
    }

    public In transition(Set<Transformer> tranSet) {
        this.trans = requireTransInTypeSame(requireNonNullElem(tranSet));
        return this;
    }

    void fillValues(Out out) {
        out.keysDef().stream().filter(keys::contains).forEach(
                key -> out.put(key.key, loadValue(key.key))
        );
    }

    protected abstract Object loadValue(String key);

    ////////////////////////////////////////////////////////////////////////////
    //***************************** static tools *****************************//

    public static In map(String key, Object value) {
        return map(Collections.singletonMap(key, value));
    }

    public static In map(Map<String, Object> map) {
        return new M(generate(map), map);
    }

    public static Builder add(String key, Object value) {
        return new Builder().add(key, value);
    }

    public static class Builder {
        private Map<String, Object> map;
        private Helper.Transformers.Builder tb;

        private Builder() {
        }

        public Builder add(String key, Object value) {
            if (map == null) map = new HashMap<>();
            map.put(key, value);
            return this;
        }

        public Builder add(Transformer trans) {
            if (tb == null) tb = Helper.Transformers.add(trans);
            else tb.add(trans);
            return this;
        }

        public In ok() {
            return map(map).transition(tb == null ? null : tb.ok());
        }
    }

    private static Set<Key$> generate(Map<String, Object> map) {
        final Set<Key$> set = new HashSet<>();
        for (Map.Entry<String, Object> entry : map.entrySet()) {
            set.add(new Key$(entry.getKey(), entry.getValue().getClass()) {
            });
        }
        return set;
    }

    private static class M extends In {
        private final Map<String, Object> map;

        private M(Set<Key$> keys, Map<String, Object> map) {
            super(keys);
            this.map = map;
        }

        @Override
        protected final Object loadValue(String key) {
            return map.get(key);
        }
    }

    public static In empty() {
        return Locker.lazyGetr(
                () -> getRef(sEmptyRef),
                () -> {
                    final In in = new In(Helper.Keys.empty()) {
                        @Override
                        void fillValues(Out out) {
                        }

                        @Override
                        protected Object loadValue(String key) {
                            return null;
                        }
                    };
                    sEmptyRef = new WeakReference<>(in);
                    return in;
                }, Locker.getLockr(In.class));
    }
}

File: src/main/scala/hobby/wei/c/reflow/In.scala
/*
 * Copyright (C) 2016-present, Wei Chou(weichou2010@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package hobby.wei.c.reflow

import hobby.chenai.nakam.lang.J2S
import hobby.chenai.nakam.lang.TypeBring.AsIs
import hobby.wei.c.reflow.Assist._
import hobby.wei.c.reflow.Dependency.MapTo
import hobby.wei.c.tool.Locker

import scala.collection._
import scala.ref.WeakReference

/**
  * @author Wei Chou(weichou2010@gmail.com)
  * @version 1.0, 14/08/2016;
  *          2.0, 23/04/2019, 修复 Scala 枚举在`In`中的 Bug。
  */
abstract class In protected (_keys: Set[KvTpe[_ <: AnyRef]], _trans: Transformer[_ <: AnyRef, _ <: AnyRef]*) {
  private[reflow] val keys: immutable.Set[KvTpe[_ <: AnyRef]]                     = requireKkDiff(requireElemNonNull(_keys.toSet))
  private[reflow] val trans: immutable.Set[Transformer[_ <: AnyRef, _ <: AnyRef]] = requireTransInTpeSame$OutKDiff(requireElemNonNull(_trans.toSet))

  private[reflow] def fillValues(out: Out): Unit = (out.keysDef & keys).foreach { key => out.put(key.key, loadValue(key.key).orNull) }

  protected def loadValue(key: String): Option[Any]
}

object In {

  def map(map: Map[String, Any], mapKes: Map[KvTpe[_ <: AnyRef], AnyRef], trans: Transformer[_ <: AnyRef, _ <: AnyRef]*): In = new M(
    generate(map) ++ mapKes.keySet,
    (map.mutable /: mapKes) { (m, kv) => m += (kv._1.key, kv._2) },
    trans: _*
  )

  def from(input: Out): In = new M(input._keys.values.toSet, input._map)

  def +[T <: AnyRef](kv: (KvTpe[T], T)): Builder = new Builder + kv

  @deprecated("Use with caution, since there is no type safety protection.")
  def +(key: String, value: Any): Builder = new Builder + (key, value)

  class Builder private[reflow] () {
    private lazy val mapKes                          = new mutable.AnyRefMap[KvTpe[_ <: AnyRef], AnyRef]
    private lazy val map                             = new mutable.AnyRefMap[String, Any]
    private lazy val tb: Helper.Transformers.Builder = new Helper.Transformers.Builder

    def +[T <: AnyRef](kv: (KvTpe[T], T)): this.type = {
      require(!map.contains(kv._1.key))
      mapKes += kv
      this
    }

    @deprecated("Use with caution, since there is no type safety protection.")
    def +(key: String, value: Any): this.type = {
      require(!mapKes.exists(_._1.key == key))
      map += (key, value)
      this
    }

    def +(ts: Transformer[_ <: AnyRef, _ <: AnyRef]*): this.type = {
      tb + (ts: _*)
      this
    }

    def ok(): In = In.map(map, mapKes, tb.ok().toSeq: _*)
  }

  private def generate(map: Map[String, Any]): Set[KvTpe[_ <: AnyRef]] =
    if (map.isEmpty) Set.empty
    else
      (new mutable.HashSet[KvTpe[_ <: AnyRef]] /: map) { (set, kv) =>
        set += new KvTpe(
          kv._1,
          kv._2.as[AnyRef] /*强制转换不会改变实际类型*/ .getClass
        )
      }

  private class M private[reflow] (keys: Set[KvTpe[_ <: AnyRef]], map: Map[String, Any], trans: Transformer[_ <: AnyRef, _ <: AnyRef]*) extends In(keys, trans: _*) {
    override protected def loadValue(key: String) = map.get(key)
  }

  def empty(): In = Locker.lazyGetr(J2S.getRef(emptyRef).orNull) {
    val in = new In(Helper.KvTpes.empty()) {
      override private[reflow] def fillValues(out: Out): Unit = {}

      override protected def loadValue(key: String) = None
    }
    emptyRef = new WeakReference(in)
    in
  }(Locker.getLockr(this))

  private var emptyRef: WeakReference[In] = _
}
